firedrake/matrix.py

```diff
@@ -34,9 +34,9 @@ def __init__(self, a, bcs, mat_type):
         # (so we can't use a set, since the iteration order may differ
         # on different processes)
 
-        ufl.Matrix.__init__(self, test.function_space(), trial.function_space())
-        # Define arguments after `Matrix.__init__` since BaseForm sets `self._arguments` to None
-        self._arguments = arguments
+        ufl.Matrix.__init__(self, test.ufl_function_space(), trial.ufl_function_space())
+        # NOTE: Do not use _arguments as the name for this variable since it is reserved by UFL
+        self._firedrake_arguments = arguments
```

Comment:
This might suggest that we want to pass `Argument`s to `ufl.Matrix` instead of `FunctionSpace`s, but until we update our UFL fork, a better fix probably is to overwrite `_analyze_form_arguments()` method as:
```
    def _analyze_form_arguments(self):
        self._arguments = (firedrake.Argument(self._ufl_function_spaces[0], 0),
                           firedrake.Argument(self._ufl_function_spaces[1], 1))
        self._coefficients = ()
```
Then, we no longer need to overwrite `arguments()` method here.

Reply:
As in you think the UFL API for `Matrix` should be changed to take arguments instead of function spaces?

Reply:
More radical counter suggestion: I don't see why we need to have our own `Argument` class. Why don't we just remove it? It seems like all of the new bits we add to `Argument` (e.g. `cell_node_map`) are just passed through from the function space anyway (e.g. https://github.com/firedrakeproject/firedrake/blob/master/firedrake/ufl_expr.py#L54) so we can just replace `arg.cell_node_map` with `arg.ufl_function_space().cell_node_map` where necessary.

As an aside this exposes a general design flaw in UFL, which I think is generally referred to as the "expression problem". If UFL exposes algorithms and abstract interfaces then things are fine, but as soon as UFL instantiates its own non-subclassed objects then we get weird inconsistencies. This is the same reason why we need `expr.subfunctions` as well as `split(expr)`. The latter returns an internal UFL `Indexed` object that we cannot intercept and replace with our own class.